fedimint-server/src/consensus/mod.rs

```diff
@@ -286,67 +279,65 @@ impl FedimintConsensus {
                 // in_scope to make sure that no await is in the middle of the span
                 let _enter = span.in_scope(|| {
                     trace!(?transaction);
-                    batch_tx.append_maybe_delete(ProposedTransactionKey(transaction.tx_hash()));
+                    dbtx.maybe_remove_entry(&ProposedTransactionKey(transaction.tx_hash()))
+                        .expect("DB Error");
 
                     // TODO: use borrowed transaction
-                    match self.process_transaction(
-                        batch_tx.subtransaction(),
-                        transaction.clone(),
-                        &caches,
-                    ) {
+                    match self.process_transaction(&mut dbtx, transaction.clone(), &caches) {
                         Ok(()) => {
-                            batch_tx.append_insert(
-                                AcceptedTransactionKey(transaction.tx_hash()),
-                                AcceptedTransaction { epoch, transaction },
-                            );
+                            dbtx.insert_entry(
+                                &AcceptedTransactionKey(transaction.tx_hash()),
+                                &AcceptedTransaction { epoch, transaction },
+                            )
+                            .expect("DB Error");
                         }
                         Err(error) => {
+                            dbtx.rollback_tx_to_savepoint();
```

Comment:
I see that rolling back in this loop, but I don't see creating a snapshot, which seems :sus:.

So this is the prod code where we need snapshots? We have bunch of txes, and if any of them is wrong, we have to roll back all the changes that one tx made, but keep going and only commit to the database the final result?

I see... well... I guess we need it then.

Reply:
Hmmmmmmmmm...

I really wish we could avoid relying on snapshot level isolation. We don't seem to be implementing it for in-memory and sled impls, and it will tie us to the rocksdb for anything serious. 

Would it be practical, to abort the whole database transaction on a invalid consensus item, mark it as invalid somewhere and try a new database transaction, each time skipping the invalid item? Or would it be too much overhead and we need it for the implementation to be practical?

Do we suspect we are going to need snapshot level isolation in other places?

Reply:
As we discussed offline, we should try to structure the code such that fedimintd is the only code that is reliant on database transactions (which most database implementations should support). Right now, none of the code relies on setting a savepoint in the middle of the transaction (which only rocksdb supports), and we should strive to keep it this way.

We will keep set_tx_savepoint exposed in the trait. I think later if we find that none of the fedimintd code needs it (and relying on database transactions is sufficient), we can remove it.